### PR TITLE
improve metadata load performance

### LINKED ISSUE #11718: SchemaBuilder#build performance improvement
### Is your feature request related to a problem?
Yes, some users have reported the metadata loading has performance issues.

### Background
Some users reported that the metadata loading performance is poor.
So I analyzed the MetaDataContextBuilder#build method's time-consuming.
First, I have this config
```rules:
- !SHARDING
  tables:
    t_order:
      actualDataNodes: ds_${0..1}.t_order_${0..1}
      tableStrategy:
        standard:
          shardingColumn: order_id
          shardingAlgorithmName: t_order_inline
      keyGenerateStrategy:
        column: order_id
        keyGeneratorName: snowflake
    t_order_item:
      actualDataNodes: ds_${0..1}.t_order_item_${0..1}
      tableStrategy:
        standard:
          shardingColumn: order_id
          shardingAlgorithmName: t_order_item_inline
      keyGenerateStrategy:
        column: order_item_id
        keyGeneratorName: snowflake
```
When proxy start. I got this time-consuming.
```
# Generate rule object
ShardingSphereRulesBuilder.buildSchemaRules time:552
# Load table metadata
SchemaBuilder.build time:109
optimizeContextFactory time:1426
# Total time-consuming method
MetaDataContextsBuilder#build time:2145
```
Then, I add 100 single tables.And not change the rule config. I got this time-consuming. Seems the time-consuming is not increased.
```
ShardingSphereRulesBuilder.buildSchemaRules time:571
SchemaBuilder.build time:131
optimizeContextFactory time:1322
MetaDataContextsBuilder#build time:2074
```
Then, I make these single tables have rules. Like this.
```
t_user_item_0:
  actualDataNodes: ds_0.t_user_item_0
t_user_item_1:
  actualDataNodes: ds_0.t_user_item_1
t_user_item_2:
  actualDataNodes: ds_0.t_user_item_2
t_user_item_3:
  actualDataNodes: ds_0.t_user_item_3
t_user_item_4:
  actualDataNodes: ds_0.t_user_item_4
t_user_item_5:
  actualDataNodes: ds_0.t_user_item_5
t_user_item_6:
  actualDataNodes: ds_0.t_user_item_6
t_user_item_7:
  actualDataNodes: ds_0.t_user_item_7
t_user_item_8:
  actualDataNodes: ds_0.t_user_item_8
t_user_item_9:
  actualDataNodes: ds_0.t_user_item_9
  ...
```
I got this time-consuming.
```
ShardingSphereRulesBuilder.buildSchemaRules time:1597
SchemaBuilder.build time:1670
optimizeContextFactory time:1474
MetaDataContextsBuilder#build time:4793
```
We can see `SchemaBuilder.build` and `ShardingSphereRulesBuilder.buildSchemaRules` time-consuming obviously increased.
Through code analysis, these 100 tables are loaded separately because they are configured with rules. This is why the time-consuming increase.
The essential reason for the slower metadata loading performance is that the rule tables are loaded one by one, and there are multiple IO operations in the process of loading each table.
### Program
```private static Map<String, TableMetaData> buildActualTableMetaDataMap(final SchemaBuilderMaterials materials) throws SQLException {
    Map<String, TableMetaData> result = new HashMap<>(materials.getRules().size(), 1);
    // replace to RuleBasedTableMetaDataBuilder's singleTable implement
    appendRemainTables(materials, result);
    // change to multi thread load and batch tables and dialect load
    for (ShardingSphereRule rule : materials.getRules()) {
        if (rule instanceof TableContainedRule) {
            for (String table : ((TableContainedRule) rule).getTables()) {
                if (!result.containsKey(table)) {
                    TableMetaDataBuilder.load(table, materials).map(optional -> result.put(table, optional));
                }
            }
        }
    }
    return result;
}
```
- Load single table by RuleBasedTableMetaDataBuilder's single table implement.
- Load rule table by using multi thread, batch load and using dialect load to decrease the IO operations.
### Task

- [ ] Load single table by RuleBasedTableMetaDataBuilder's single table implement.

- [ ] Load rule table by using multi thread, batch load and using dialect load.

Thanks for @Beyondeclipse ideas and suggestions

### PR DESCRIPTION
**Pull request by tuichenchuxin**
Fixes #11718 .

Changes proposed in this pull request:
- SchemaBuilder#build improve performance
- add test case

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataBuilder.java [hunk 1]
```diff
@@ -79,7 +83,36 @@
...
+    /**
+     * Load logic table metadata.
+     *
+     * @param materials schema builder materials
+     * @param executorService executorService
+     * @return table meta data collection
+     * @throws SQLException SQL exception
+     */
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    public static Optional<Collection<TableMetaData>> loadLogicTables(final SchemaBuilderMaterials materials, final ExecutorService executorService) throws SQLException {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataBuilder.java:96]: @tuichenchuxin Why add Optional for Collection result?

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java [hunk 2]
```diff
@@ -49,6 +60,73 @@
...
+            throw new ShardingSphereException(ex);
+        }
+        return result;
+    }
+    
+    private Map<String, Collection<String>> getDataSourceTablesGroup(final Collection<String> tableNames, final Map<String, DataSource> dataSourceMap, final DataNodes dataNodes) {
+        Map<String, Collection<String>> result = new LinkedHashMap<>();
+        for (String tableName : tableNames) {
+            String dataSourceName = dataNodes.getDataNodes(tableName).stream().map(DataNode::getDataSourceName).findFirst().orElseGet(() -> dataSourceMap.keySet().iterator().next());
+            Collection<String> collection = result.getOrDefault(dataSourceName, new LinkedList<>());
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java:102]: @tuichenchuxin `tableNames` is better.

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java [hunk 3]
```diff
@@ -49,6 +60,73 @@
...
+            collection.add(tableName);
+            result.put(dataSourceName, collection);
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java:104]: @tuichenchuxin Use `putIfAbsent` is better.

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java [hunk 4]
```diff
@@ -49,6 +61,72 @@
                 ? TableMetaDataLoader.load(dataSourceMap.get(dataSourceName), tableName, databaseType) : Optional.empty();
     }
     
+    @Override
+    public Map<String, TableMetaData> load(final Collection<String> tableNames, final DatabaseType databaseType, final Map<String, DataSource> dataSourceMap, final DataNodes dataNodes,
```
Comment (taojintianxia) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java:65]: too much parameters, if possible, why not encapsulate as a bean ?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/H2TableMetaDataLoader.java [hunk 5]
```diff
@@ -164,14 +178,15 @@ private String getPrimaryKeyMetaDataSQL(final Collection<String> existedTables)
         return result;
     }
 
-    private String getGeneratedInfoSQL(final Collection<String> existedTables) {
-        return existedTables.isEmpty() ? GENERATED_INFO_SQL
-                : String.format(GENERATED_INFO_SQL_WITH_EXISTED_TABLES, existedTables.stream().map(each -> String.format("'%s'", each)).collect(Collectors.joining(",")));
+    private String getGeneratedInfoSQL(final Collection<String> tables, final boolean isExclude) {
```
Comment (taojintianxia) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/H2TableMetaDataLoader.java:181]: conditional-operator appear two times in one clause ? I believe this is easy to make people confused.

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java (hunk 4 above)
Comment (tuichenchuxin) [reply shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java:65]: ok，I'll try to decrease parameters.

**Review by tristaZero — CHANGES_REQUESTED** [review]
Do you think `loadByDialect` can be extracted into its father class?

**Review by tristaZero — CHANGES_REQUESTED** [review]
What's the function of `isExclude`?

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java [hunk 6]
```diff
@@ -49,6 +57,31 @@
...
+        }
+        Map<String, Collection<String>> dataSourceTables = getTableGroup(loadTableNames, materials);
+        return loader.isPresent() ? TableMetaDataLoader.load(loader.get(), dataSourceTables, materials.getDataSourceMap(), executorService)
+                : TableMetaDataLoader.load(dataSourceTables, materials.getDatabaseType(), materials.getDataSourceMap());
+    }
+    
+    private Map<String, Collection<String>> getTableGroup(final Collection<String> tableNames, final SchemaBuilderMaterials materials) {
+        Map<String, Collection<String>> result = new LinkedHashMap<>();
+        DataNodes dataNodes = new DataNodes(materials.getRules());
+        for (String tableName : tableNames) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/metadata/EncryptTableMetaDataBuilder.java:76]: @tuichenchuxin modify `tableName` to `each`.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/metadata/ShardingTableMetaDataBuilder.java [hunk 7]
```diff
@@ -82,6 +87,63 @@
...
+            return Collections.emptyMap();
+        }
+        boolean isCheckingMetaData = materials.getProps().getValue(ConfigurationPropertyKey.CHECK_TABLE_METADATA_ENABLED);
+        return isCheckingMetaData ? loadWithCheck(loadTableNames, rule, materials) : loadWithOutCheck(loadTableNames, rule, materials, executorService);
+    }
+    
+    private Map<String, TableMetaData> loadWithCheck(final Collection<String> tableNames, final ShardingRule rule, final SchemaBuilderMaterials materials) {
+        int maxConnectionsSizePerQuery = materials.getProps().getValue(ConfigurationPropertyKey.MAX_CONNECTIONS_SIZE_PER_QUERY);
+        Map<String, TableMetaData> result = new HashMap<>();
+        for (String tableName : tableNames) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/metadata/ShardingTableMetaDataBuilder.java:104]: @tuichenchuxin Same problem.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/metadata/ShardingTableMetaDataBuilder.java [hunk 8]
```diff
@@ -82,6 +87,63 @@
...
+            result.put(tableRule.getLogicTable(), actualTableMetaDataMap.values().iterator().next());
+        }
+        return result;
+    }
+    
+    private Map<String, TableMetaData> loadWithOutCheck(final Collection<String> tableNames, final ShardingRule rule,
+                                                        final SchemaBuilderMaterials materials, final ExecutorService executorService) throws SQLException {
+        Optional<DialectTableMetaDataLoader> loader = TableMetaDataLoader.findDialectTableMetaDataLoader(materials.getDatabaseType());
+        Map<String, Collection<String>> dataSourceTables = getTableGroup(tableNames, materials);
+        return loader.isPresent() ? new LinkedHashMap<>(getLogicTableName(TableMetaDataLoader.load(loader.get(), dataSourceTables, materials.getDataSourceMap(), executorService).values(), rule))
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/metadata/ShardingTableMetaDataBuilder.java:121]: @tuichenchuxin Is it better to get the results first and then decorate them?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/metadata/ShardingTableMetaDataBuilder.java [hunk 9]
```diff
@@ -82,6 +87,63 @@
...
+            tables.add(dataNode.getTableName());
+            result.putIfAbsent(dataNode.getDataSourceName(), tables);
+        }
+        return result;
+    }
+    
+    private Map<String, TableMetaData> getLogicTableName(final Collection<TableMetaData> tableMetaDatas, final ShardingRule rule) {
+        Map<String, TableMetaData> result = new LinkedHashMap<>();
+        for (TableMetaData each : tableMetaDatas) {
+            if (rule.findLogicTableByActualTable(each.getName()).isPresent()) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/metadata/ShardingTableMetaDataBuilder.java:140]: @tuichenchuxin Please avoid multiple calls `findLogicTableByActualTable` method, using `ifPresent`?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/TableMetaDataLoader.java [hunk 10]
```diff
@@ -59,6 +69,26 @@
...
+     * @param dataSourceTable data source table name map
+     * @param databaseType database type
+     * @param dataSourceMap data source map
+     * @return table meta data map
+     * @throws SQLException SQL exception
+     */
+    public static Map<String, TableMetaData> load(final Map<String, Collection<String>> dataSourceTable, final DatabaseType databaseType,
+                                                  final Map<String, DataSource> dataSourceMap) throws SQLException {
+        Map<String, TableMetaData> result = new LinkedHashMap<>();
+        for (Entry<String, Collection<String>> each : dataSourceTable.entrySet()) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/TableMetaDataLoader.java:84]: @tuichenchuxin `entry` is better.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/TableMetaDataLoader.java [hunk 11]
```diff
@@ -59,6 +69,26 @@
...
+            for (String tableName : each.getValue()) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/TableMetaDataLoader.java:85]: @tuichenchuxin `each` is better.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataBuilder.java [hunk 12]
```diff
@@ -79,7 +83,40 @@
...
+        return result;
+    }
+    
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private static void loadTableContainedRuleTables(final SchemaBuilderMaterials materials, final ExecutorService executorService, final Collection<TableMetaData> result,
+                                                     final Entry<ShardingSphereRule, RuleBasedTableMetaDataBuilder> ruleBuilderEntry) throws SQLException {
+        TableContainedRule rule = (TableContainedRule) ruleBuilderEntry.getKey();
+        RuleBasedTableMetaDataBuilder loader = ruleBuilderEntry.getValue();
+        Collection<String> needLoadTables = rule.getTables().stream()
+                .filter(each -> !result.stream().map(TableMetaData::getName).collect(Collectors.toList()).contains(each)).collect(Collectors.toList());
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataBuilder.java:112]: @tuichenchuxin Please extract this logic to improve performance.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Review by strongduanmu — APPROVED** [review]

**Review by tristaZero — APPROVED** [review]

**Pull request merged**